### PR TITLE
Enable the IN operator to also compare the value of single-pair structs (Revised)

### LINKED ISSUE #524: SQL IN Query is not working as expected
Hello, we have the following query with IN clause that is returning empty results:

```
SELECT * FROM users WHERE name IN (SELECT name FROM roles)
```

However, the same query works in Postgres.  Alternatively following works, but of course, it's not same query semantically:

```
SELECT * FROM users u WHERE EXISTS (SELECT r.name FROM roles r WHERE r.name = u.name)
```

What are we doing/assuming that is wrong?

Sample code to make trying easier:

```kotlin
  val ion = IonSystemBuilder.standard().build()
  val pipeline = CompilerPipeline.standard(ion)

  val globalVariables = Bindings.buildLazyBindings<ExprValue> {
    addBinding("users") {
      val values = ion.loader.load(
        """
        {
        "name": "mustafa",
        "age": 30,
        }
        {
        "name": "ilteris",
        "age": 26,
        }
        {
        "name": "onur",
        "age": 24,
        }""".trimMargin()
      )
      pipeline.valueFactory.newFromIonValue(values)
    }
    addBinding("roles") {
      val values = ion.loader.load(
        """
        {
        "username": "mustafa",
        "admin": true,
        }
        {
        "name": "ilteris",
        "admin": false,
        }""".trimMargin()
      )
      pipeline.valueFactory.newFromIonValue(values)
    }
  }

  val query = "SELECT * FROM users WHERE name IN (SELECT name FROM roles)"
  val e = pipeline.compile(query)
  val session = EvaluationSession.build {
    globals(globalVariables)
  }

  val result = e.eval(session)
  for (exprValue in result) {
    println(exprValue)
  }
```

### PR DESCRIPTION
**Pull request by johnedquinn**
### Related Issue
Resolves #524 

### Note
I actually made a PR #614, but I merged all of @dlurton's changes in, and it made it impossible to read. This is a new PR, and I will be closing/cancelling the other.

## Description
Allows the IN clause to compare the value of single-pair structs, to maintain compatibility with SQL-92. To do this, it is necessary to make two changes:
- If the rightOp is already computed as a sequence of single-pair structs (where each pair's value is a literal), we can use the optimized approach and immediately compare with that value.
- If the rightOp is not computed (for example: another SELECT returning a bag/list of structs, or a bag of structs with pairs that contain un-evaluated values), then we need to evaluate.

### Examples

```sql
SELECT age FROM << { 'name': 'John', 'age': 22 } >> WHERE name IN (SELECT name FROM <<{ 'name': 'John' }>>)
```
should result in:
```ion
<<
  { age: 22 }
>>
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt
Comment (lziq) [inline lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:827]:
```suggestion
                                    ExprValueType.STRUCT -> if (it.ionValue.asIonStruct().size() == 1 && it.iterator().next().exprEquals(leftValue)) {
                                        return@thunkEnvOperands valueFactory.newBoolean(true)
                                    }
```

**Comment by lziq** [comment]
And it seems there are some merge conflicts. I think it will be good to go after you resolve them.

**Review by lziq — APPROVED** [review]

**Comment by johnedquinn** [comment]
I'm currently holding off on merging this PR for one reason. If you take a look at the linked issue, their failing query is:

```sql
SELECT * FROM users WHERE name IN (SELECT name FROM roles)
```

Their sub-query is returning a single-column table (bag) with the attribute `name`. They are able to use the main query's variable `name` (a literal -- likely string) to compare the single-column table with the IN clause.

The same query in PartiQL does NOT work. In PartiQL, the sub-query returns a bag of structs (tuples). Something like:
```partiql
<< { 'name': 'John' }, { 'name': 'Ralph' } >>
```

So, when the main query variable `name` attempts to compare to the sub-query result (a bag of structs, not a bag of strings), there is a type mismatch. While we can modify the IN clause to compare the left operand (in this case, a String) with the value of the key-value pair found in the right operand, it might do us harm.

Essentially, what if a user wanted to perform the following:

```sql
SELECT * FROM << { 'name': 'John', 'age': 22 } >> WHERE { 'name': name } IN (SELECT name FROM << { 'name': 'John' } >>)
```

This query explicity wants a comparison between a struct `{ 'name': name }` and a bag of structs `<< { 'name': 'John' } >>`. If we add a work-around, to recognize when the right operand is a bag of single-value structs, then we might be limiting the end-user when they DO in fact want to compare structs.

Need to work with the team to see what we want to do.

**Comment by johnedquinn** [comment]
See another example:

```partiql
users = << { 'name': 'Ziqi Li', 'age': '22' }, { 'name': 'Ziqi Li', 'age': '23'} >>
info = << { 'name': 'Ziqi Li', 'age': '22', 'height': '5.10', 'addr': '1234 Terry Ave' } >>
```

```sql
SELECT * FROM info WHERE { 'name': name, 'age': age } IN (SELECT name, age FROM users)
```

**Comment by am357** [comment]
> I'm currently holding off on merging this PR for one reason. If you take a look at the linked issue, their failing query is...

Glad you've brought it up. My reading from [spec.'s section 9.1](https://partiql.org/assets/PartiQL-Specification.pdf#subsection.9.1) is that we should not coerce the subquery expression if its the `rhs` of `IN`:

---
In each of the following cases a SFW subquery coerces into a scalar:
- if it appears as the rhs of a comparison operator (=, >, etc) where the lhs is not an array literal. And, vice
versa, if it appears as the lhs of a comparison operator where the rhs is not an array literal. (If it is the lhs of a
comparison operator where the lhs is an array literal, it coerces into array, per Section 9.2.)
- if it is an SFW subquery expression that (a) is not the collection expression of a FROM clause item and **(b) is not
the rhs of an IN. (If it is the rhs of an IN then it should not be coerced;** see note on semantics of IN, Section ??.)
---

Considering above, I also share your concern on loosing the ability of comparing structs. Let's have a further discussion w/ other maintainers.

**Comment by johnedquinn** [comment]
Thanks for referencing the respective section in the spec ! @am357

**Pull request closed without merging**
